Sources/AsyncDNSResolver/dnssd/DNSResolver_dnssd.swift

```diff
@@ -158,9 +162,32 @@ struct DNSSD {
                 return continuation.finish(throwing: AsyncDNSResolver.Error(dnssdCode: _code))
             }
 
+            let serviceSockFD = DNSServiceRefSockFD(serviceRefPtr.pointee)
+            guard serviceSockFD != -1 else {
+                return continuation.finish(throwing: AsyncDNSResolver.Error(code: .internalError, message: "Failed to access the DNSSD service socket"))
+            }
+
+            var pollFDs = [pollfd(fd: serviceSockFD, events: Int16(POLLIN), revents: 0)]
```

Comment:
`poll` is blocking until this is ready so we can't do this here. If you want async I/O you'd need to use SwiftNIO here.

Reply:
We did work to remove the SwiftNIO dependency earlier (https://github.com/apple/swift-async-dns-resolver/pull/20), so I don't think we want to add it back.

The intention here is to allow checking for task cancellation, because calling `DNSServiceProcessResult` alone just blocks and doesn't seem to give us the chance to do that at all.

The docs for `DNSServiceProcessResult ` says:

```
This call will block until the daemon's response is received. Use DNSServiceRefSockFD() in
conjunction with a run loop or select() to determine the presence of a response from the
server before calling this function to process the reply without blocking.
```

That's why we went with this approach here. It doesn't sound like we must do async I/O to see improvement, but I could be wrong of course.

Reply:
If `poll` isn't called with a timeout, it'll block. If we mustn't use NIO, prefer `DispatchSource`s.

Reply:
> Use DNSServiceRefSockFD() in conjunction with a run loop 

^^ that's exactly what NIO can do for you.

What's the goal with removing the NIO dependency? Every adopter will have a NIO dependency anyway, right?

Reply:
> What's the goal with removing the NIO dependency? 

Only a couple of methods on `ByteBuffer` were being used for parsing which was quite heavy handed and replaced with `ArraySlice<UInt8>`. There's no requirement to be free of NIO.

> Every adopter will have a NIO dependency anyway, right? 

Not necessarily although I'm sure many will.

Reply:
> If poll isn't called with a timeout, it'll block.

We are passing `0` as timeout I believe ([`poll(&pollFDs, 1, 0)`](https://github.com/apple/swift-async-dns-resolver/pull/34/files#diff-a647939064e83990e89a15c5f7e4b4b114e1cb0c3a9299cac47f79fa5a3b8462R176)), so it should return immediately according to the [man page](https://linux.die.net/man/2/poll):

```
Specifying a negative value in timeout means an infinite timeout.  
Specifying a timeout of zero causes poll() to return immediately, 
even if no file descriptors are ready.
```

I wonder if it should sleep a bit before calling `poll` again though.

> that's exactly what NIO can do for you

Is there an example that we can follow? I found these threads:
- https://forums.swift.org/t/socket-library-built-with-swift-concurrency/56505
- https://github.com/apple/swift-nio/issues/2074

But I didn't find anything obvious.

I don't doubt that SwiftNIO can do this more efficiently, but at the end of the day we need to look at the trade-offs. Does doing it without SwiftNIO give us a "good enough" solution? Is it worthwhile to add SwiftNIO, which is a heavier dependency, for this? (I don't have problem with adding the SwiftNIO dependency FWIW.)

Reply:
Hot-looping on poll is not the right choice: it's not meaningfully different than just sleeping the thread, as we'll never yield it.

Doing this with NIO directly is gonna be a little painful as you'd have to write a `Channel` to intercept the `readFromSocket` function. Not the end of the world, but not as easy as you might like. I continue to think that DispatchSource is going to be the easiest way to achieve this goal.

Reply:
Oh yes, if you don't actually need to do anything with the data in there, then DispacthSource is good. It's pretty bad if you need to actually read the data but if you just need to be told that there is something to read, `DispatchSource.makeReadSource` is fine

Reply:
Thanks for your feedback, I'm going to try implementing this with DispatchSource then, if it's ok for everyone.